.github/workflows/android-perf.yml

```diff
@@ -0,0 +1,179 @@
+
+        # TODO: This needs to be replaced with a generic loader .apk
+        # Build LLM Demo for Android
+        bash build/build_android_llm_demo.sh ${{ matrix.tokenizer }} ${ARTIFACTS_DIR_NAME}
+  # Upload artifacts to S3. The artifacts are needed not only by the device farm but also TorchChat
+  upload-android-apps:
+    needs: build-llm-demo
+    runs-on: linux.2xlarge
+    steps:
+      - name: Download the artifacts from GitHub
+        uses: actions/download-artifact@v3
+        with:
+          # The name here needs to match the name of the upload-artifact parameter
+          name: android-apps
+          path: ${{ runner.temp }}/artifacts/
+
+      - name: Verify the artifacts
+        shell: bash
+        working-directory: ${{ runner.temp }}/artifacts/
+        run: |
+          ls -lah ./
+
+      - name: Upload the artifacts to S3
+        uses: seemethere/upload-artifact-s3@v5
+        with:
+          s3-bucket: gha-artifacts
+          s3-prefix: |
+            ${{ github.repository }}/${{ github.run_id }}/artifact
+          retention-days: 14
+          if-no-files-found: ignore
+          path: ${{ runner.temp }}/artifacts/
+
+  # Let's see how expensive this job is, we might want to tone it down by running it periodically
+  benchmark-on-device:
+    needs:
+      - upload-models
+      - upload-android-apps
+    permissions:
+      id-token: write
+      contents: read
+    uses: pytorch/test-infra/.github/workflows/mobile_job.yml@main
+    strategy:
+      matrix:
+        # https://github.com/pytorch/executorch/blob/main/examples/demo-apps/android/LlamaDemo/README.md#alternative-2-build-from-local-machine
+        # mentions that tiktoken is only for Llama3. So, we can export it later in another archive
+        # like https://ossci-assets.s3.amazonaws.com/executorch-android-llama2-7b-0717.zip when this is
+        # updated to run Llama3
+        tokenizer: [bpe]
+    with:
+      device-type: android
+      runner: linux.2xlarge
+      test-infra-ref: ''
+      # This is the ARN of ExecuTorch project on AWS
+      project-arn: arn:aws:devicefarm:us-west-2:308535385114:project:02a2cf0f-6d9b-45ee-ba1a-a086587469e6
+      # This is the custom Android device pool that only includes Samsung Galaxy S2x
+      device-pool-arn: arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2cf0f-6d9b-45ee-ba1a-a086587469e6/e59f866a-30aa-4aa1-87b7-4510e5820dfa
+      # Uploaded to S3 from the previous job, the name of the app comes from the project itself
+      android-app-archive: https://gha-artifacts.s3.amazonaws.com/${{ github.repository }}/${{ github.run_id }}/artifact/llm_demo_${{ matrix.tokenizer }}/app-debug.apk
+      android-test-archive: https://gha-artifacts.s3.amazonaws.com/${{ github.repository }}/${{ github.run_id }}/artifact/llm_demo_${{ matrix.tokenizer }}/app-debug-androidTest.apk
+      # The test spec can be downloaded from https://ossci-assets.s3.amazonaws.com/android-llama2-device-farm-test-spec.yml
```

Comment:
@huydhn This test spec is llama2 specific in which the model and its tokenizer are hard-coded. Should we create more test specs for different models/apps (less preferred) or could we make it model/app agnostic so that we can reuse it for general purpose?

Reply:
As I can see it, as long as we have a model/app agnostic script to run the test, the script can be adopted to the test spec format.  Another thing is to figure out how to send the results back in a generic way.  Atm, that depends on parsing `[PyTorch]` from the instrumentation test log, but I think, in the context of perf testing, being able to dump the results into a CSV/JSON file, then copy it from the devices is better.

Reply:
> As I can see it, as long as we have a model/app agnostic script to run the test, the script can be adopted to the test spec format, giving us a model/app agnostic spec.

@huydhn Okay. This sounds flexible as we the only thing needs to be hard-coded in the test-spec is the location of the script. However, to test the workflow, I still need a way to test non-llama2 models. Right now the test spec is hard-coded with llama2: https://www.internalfb.com/phabricator/paste/view/P1503784701?lines=11-18

> Another thing is to figure out how to send the results back in a generic way. Atm, that depends on parsing [PyTorch] from the instrumentation test log, but I think, in the context of perf testing, being able to dump the results into a CSV/JSON file, then copy that file back from the devices is better.

I think we can reuse AIBench's format to dump the benchmark result into json and having a parser for it. That can be reused. For now it's okay to parsing [PyTorch] label from the log.

Reply:
Fyi, [T197321521](https://www.internalfb.com/intern/tasks/?t=197321521) and [T197322159](https://www.internalfb.com/intern/tasks/?t=197322159)